docs/architecture/adr-002-api-cleanup.md

```diff
@@ -0,0 +1,94 @@
+# ADR ADR-002: API Cleanup to Improve Commit Performance
+
+## Changelog
+
+* 2023-06-06: First draft
+
+## Status
+
+DRAFT
+
+## Abstract
+
+This ADR proposes a cleanup of the API to make more understandable and maintainable the codebase of `iavl`.
+
+There is a lot of legacy code in the SDK that is not used anymore and can be removed. See the [Discussion](https://github.com/cosmos/iavl/issues/737) for more details. 
+
+There are some proposals for the speedup of the `Commit` by the async writes. See the [Discussion](https://github.com/cosmos/cosmos-sdk/issues/16173) for more details.
+
+## Context
+
+The current implementation of `iavl` suffers from performance issues due to synchronous writes during the `Commit` process. To address this, the proposed changes aim to finalize the current version in memory, parallelize hash calculations in `WorkingHash`, and introduce asynchronous writes in the background.
+
+It is also necessary to refactor the `Set` function to accept a batch object for the atomic commitments. 
+
+Moreover, the existing architecture of `iavl` lacks modularity and code organization:
+
+- The boundary between `ImmutableTree` and `MutableTree` is not clear.
+- There are many public methods in `nodeDB`, making it less structured.
+
+## Decision
```

Comment:
what is the external api to users? the below seems to focus on internal changes.

Reply:
do you think iavl needs the external API interface to clarify? I plan to keep the majority external API